src/Compilers/CSharp/Test/Emit2/Emit/RuntimeProbing/ModuleCancellationTests.cs

```diff
@@ -0,0 +1,2219 @@
+              // sequence point: }
+              IL_0018:  ret
+            }
+            """);
+    }
+
+    [Fact]
+    public void ArgumentReplacement_Overload_Generic_CompatibleButNotSameConstraint()
+    {
+        var source = """
+            using System.Threading;
+            using System.Threading.Tasks;
+            
+            class C
+            {
+                void F(CancellationToken token)
+                {
+                    G<int>(1);
+                }
+
+                void G<T>(int a) => throw null;            
+                void G<T>(int a, CancellationToken token) where T : struct => throw null;
+            }
+            """;
+
+        var verifier = CompileAndVerify(source);
+
+        AssertNotInstrumentedWithTokenLoad(verifier, "C.F");
+    }
+
+    [Fact]
+    public void ArgumentReplacement_Overload_Generic_IncompatibleConstraint()
+    {
+        var source = """
+            using System.Threading;
+            using System.Threading.Tasks;
+            
+            class C
+            {
+                void F(CancellationToken token)
+                {
+                    G<int>(1);
+                }
+
+                void G<T>(int a) => throw null;            
+                void G<T>(int a, CancellationToken token) where T : class => throw null;
+            }
+            """;
+
+        var verifier = CompileAndVerify(source);
+
+        AssertNotInstrumentedWithTokenLoad(verifier, "C.F");
+    }
+
+    [Fact]
+    public void ArgumentReplacement_Overload_NoMatching()
+    {
+        var source = """
+            using System.Threading;
+            using System.Threading.Tasks;
+            
+            class C
+            {
+                void F(CancellationToken token)
+                {
+                    G(1);
+                }
+
+                void G(int a)
+                {
+                }
+            
+                int G(int a, CancellationToken token) => throw null;
+                void G(out int a, CancellationToken token) => throw null;
+                void G(long a, CancellationToken token) => throw null;
+                void G(CancellationToken token) => throw null;
+                void G(CancellationToken token, int a) => throw null;
+                void G(int a, int b, CancellationToken token) => throw null;
+                void G<T>(int a, CancellationToken token) => throw null;
+            }
+            """;
+
+        var verifier = CompileAndVerify(source);
+
+        AssertNotInstrumentedWithTokenLoad(verifier, "C.F");
+    }
+
+    [Fact]
+    public void ArgumentReplacement_Overload_ReturnType()
+    {
+        var source = """
+            using System.Threading;
+            using System.Threading.Tasks;
+            
+            class C
+            {
+                void F(CancellationToken token)
+                {
+                    G(1);
+                }
+
+                int G(int a) => throw null;
```

Comment:
>int G(int a) => throw null; [](http://example.com/codeflow?start=16&length=27)

Are we testing scenario when this method returns a ref and the other one doesn't?
Also consider throwing `ref readonly` vs. `ref` vs. nothing into the mix.